crates/vm/src/wasm/host_env.rs

```diff
@@ -3,13 +3,208 @@
 //! Here, we expose the host functions into wasm's
 //! imports, so they can be called from inside the wasm.
 
+use std::cell::RefCell;
+use std::rc;
+
+use namada_core::hints;
+use namada_gas::{Gas, GasMetering};
 use namada_state::{DB, DBIter, StorageHasher};
 use wasmer::{Function, FunctionEnv, Imports};
 
 use crate::host_env::{TxVmEnv, VpEvaluator, VpVmEnv};
 use crate::wasm::memory::WasmMemory;
 use crate::{WasmCacheAccess, host_env};
 
+/// Wasm native gas meter
+#[derive(Debug, Default)]
+pub struct WasmGasMeter {
+    gas_scale: u64,
+    initial_gas: Gas,
+    tx_gas_limit: Gas,
+    wasm_transaction_gas_global: Option<wasmer::Global>,
+    store: Option<rc::Weak<RefCell<wasmer::Store>>>,
+}
+
+impl WasmGasMeter {
+    /// Create an uninitialized wasm gas meter.
+    ///
+    /// The meter will only be initialized after a wasm instance
+    /// is built, populating the expected global variable that
+    /// will track gas usage.
+    pub const fn uninit() -> Self {
+        Self {
+            gas_scale: 0u64,
+            initial_gas: Gas::new(0u64),
+            tx_gas_limit: Gas::new(0u64),
+            wasm_transaction_gas_global: None,
+            store: None,
+        }
+    }
+
+    /// Initialize the wasm gas meter.
+    pub fn init_from(
+        &mut self,
+        meter: &impl GasMetering,
+        gas_global: wasmer::Global,
+        store: rc::Weak<RefCell<wasmer::Store>>,
+    ) {
+        self.gas_scale = meter.get_gas_scale();
+        self.tx_gas_limit = meter.get_gas_limit();
+        self.initial_gas = meter.get_available_gas();
+        self.wasm_transaction_gas_global = Some(gas_global);
+        self.store = Some(store);
+
+        self.write_wasm_gas(self.initial_gas.clone(), None);
+    }
+
+    /// Return the gas consumed while executing wasm code.
+    pub fn wasm_used_gas(&self) -> namada_gas::Result<Gas> {
+        // initial  := limit - <sigchecks>
+        // variable := initial - wasm
+        //      wasm = initial - variable
+
+        let current_tx_gas = self.read_wasm_gas(None);
+
+        if current_tx_gas == u64::MAX.into() {
+            return Err(namada_gas::Error::TransactionGasExceededError(
+                self.tx_gas_limit.get_whole_gas_units(self.gas_scale),
+            ));
+        }
+
+        self.initial_gas
```

Comment:
Doesn't this return the gas left to be used instead of the gas used so far? Shouldn't the latter just be `current_tx_gas`

Reply:
Ah ok it looks like the wasm global is tracking the gas left, not the one used so far, so you need to run this subtraction